keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -0,0 +1,1354 @@
+    // matches the DriverName set in the ResourceClaimStatus this
+    // ResourceHandle is embedded in. It is set at allocation time and is
+    // intended for processing by the kubelet plugin whose name matches
+    // the DriverName set in this ResourceHandle.
+    //
+    // The maximum size of this field is 16KiB. This may get increased in the
+    // future, but not reduced.
+    // +optional
+    Data *string
+
+    // If StructuredData is set, then it needs to be used instead of Data.
+    StructuredData *StructuredResourceHandle
+}
+
+type StructuredResourceHandle struct {
+    // VendorClassParameters are the parameters from the time that the claim
+    // was allocated. They can be arbitrary setup parameters that are
+    // ignore by the counter controller.
+    VendorClassParameters *runtime.RawExtension
+
+    // VendorClaimParameters are the parameters from the time that the claim was
+    // allocated. Some of the fields were used by the counter controller to
+    // allocated resources, others can be arbitrary setup parameters.
+    VendorClaimParameters *runtime.RawExtension
+
+   	// NodeName is the name of the node providing the necessary resources.
+    // This mirrors the AllocationResult.AvailableOnNodes with a simpler
+    // type.
+    //
+    // The driver name is the one stored in ResourceHandle.
+    NodeName string
+
+    AllocationResultModel // inline, fields must not conflict
+}
+
+// AllocationResultModel must have one and only one field set.
+type AllocationResultModel struct {
+    NamedResourcesWithAttributes *NamedResourcesWithAttributesResult
+}
+
+type NamedResourcesWithAttributesResult struct {
+    Resources []AllocatedResource
+}
+
+type AllocatedResource struct {
+   ID string // A concatenation with / of the individual names.
+}
+```
+
+### Implementation of structured models
+
+In the Go types above, all structs starting with `NamedResourcesWithAttributes`
+are part of that structured model. In practice, organizing those inside their own
+Go package and then importing that package in the definition of the
+resource.k8s.io API will result in a cleaner separation at the source code
+level. It has no impact on the resulting Kubernetes API.
+
+### Scheduling + Allocation
+
+The dynamic resource scheduler plugin handles the common fields of
+NodeResourceSlice, ResourceClaimParameters and StructuredResourceHandle. For the
+structured model fields it calls out to code that is associated with the
+corresponding model.
+
+During filtering it is decided which nodes have the necessary resources. If a
+node is found, the scheduler plugin updates the resource claim status as part
+of goroutine which handles pod binding.
+
+Like a normal DRA driver controller, the scheduler also sets a finalizer to
+ensure that users cannot accidentally delete the allocated claim while a pod
+is about to start which depends on it. That finalizer is
+"structured.dra.k8s.io/delete-protection".
```

Comment:
why not just `dra.k8s.io/delete-protection`?

Reply:
That also works.